src/raft.c

```diff
@@ -1298,14 +1517,25 @@ void handleAppendEntries(RedisRaftCtx *rr, RaftReq *req)
 
 static void handleReadOnlyCommand(void *arg, int can_read)
 {
+    RedisRaftCtx *rr = &redis_raft;
     RaftReq *req = arg;
 
     if (!can_read) {
         RedisModule_ReplyWithError(req->ctx, "TIMEOUT no quorum for read");
         goto exit;
     }
 
-    executeRaftRedisCommandArray(&req->r.redis.cmds, req->ctx, req->ctx);
+    RaftRedisCommandArray * cmds = &req->r.redis.cmds;
+
+    HandleAsking(cmds);
+
+    if (rr->config->sharding && handleSharding(rr, req, cmds) != RR_OK) {
+        goto exit;
+    }
+
+    if (cmds->slot == -1 || validateRaftRedisCommandArray(&redis_raft, cmds, req->ctx) == RR_OK) {
```

Comment:
I think if you move this part info `executeRaftRedisCommandArray()`, you can avoid changes in this function, `handleReadOnlyCommand()` and in `executeLogEntry()`. Both needs to handle "asking, sharding and slot validation". If `executeRaftRedisCommandArray()` does not, we'll doing these in a single place. Just consider if it makes sense.

Reply:
ok, tried this.